neural_lam/weather_dataset.py

```diff
@@ -1,262 +1,635 @@
+                    "only using first ensemble member, so dataset size is "
+                    " effectively reduced by the number of ensemble members "
+                    f"({self.da_state.ensemble_member.size})",
+                    UserWarning,
+                )
+
+            # check that there are enough forecast steps available to create
+            # samples given the number of autoregressive steps requested
+            n_forecast_steps = self.da_state.elapsed_forecast_duration.size
+            if n_forecast_steps < 2 + self.ar_steps:
+                raise ValueError(
+                    "The number of forecast steps available "
+                    f"({n_forecast_steps}) is less than the required "
+                    f"2+ar_steps (2+{self.ar_steps}={2 + self.ar_steps}) for "
+                    "creating a sample with initial and target states."
+                )
+
+            return self.da_state.analysis_time.size
+        else:
+            # Calculate the number of samples in the dataset n_samples = total
+            # time steps - (autoregressive steps + past forcing + future
+            # forcing)
+            #:
+            # Where:
+            #   - total time steps: len(self.da_state.time)
+            #   - autoregressive steps: self.ar_steps
+            #   - past forcing: max(2, self.include_past_forcing) (at least 2
+            #     time steps are required for the initial state)
+            #   - future forcing: self.include_future_forcing
+            return (
+                len(self.da_state.time)
+                - self.ar_steps
+                - max(2, self.include_past_forcing)
+                - self.include_future_forcing
+            )
+
+    def _slice_state_time(self, da_state, idx, n_steps: int):
+        """
+        Produce a time slice of the given dataarray `da_state` (state) starting
+        at `idx` and with `n_steps` steps. An `offset`is calculated based on the
+        `include_past_forcing` class attribute. `Offset` is used to offset the
+        start of the sample, to assert that enough previous time steps are
+        available for the 2 initial states and any corresponding forcings
+        (calculated in `_slice_forcing_time`).
+
+        Parameters
+        ----------
+        da_state : xr.DataArray
+            The dataarray to slice. This is expected to have a `time` dimension
+            if the datastore is providing analysis only data, and a
+            `analysis_time` and `elapsed_forecast_duration` dimensions if the
+            datastore is providing forecast data.
+        idx : int
+            The index of the time step to start the sample from.
+        n_steps : int
+            The number of time steps to include in the sample.
+        """
+        # The current implementation requires at least 2 time steps for the
+        # initial state (see GraphCast).
+        init_steps = 2
+        offset = idx + max(init_steps, self.include_past_forcing)
```

Comment:
(Ignoring self.include_past_forcing here) Why do we have to offset this by `init_state=2`? This method returns the full sequence of states (initial conditions and targets), which are later split up. There is no other method that uses the entries at (idx-1, idx-2).

Reply:
This is actually a bug. You want the times to exactly match between `target_states` and `forcing`. I pushed a fix for this, tested on the dummy datastore: https://github.com/mllam/neural-lam/pull/66/commits/599917d79b1a54f45980ab8c48e5bea3e91bfde2

Reply:
Had to apply another bugfix here: @joeloskarsson please double-check 
[Cover cases where include_past_forcing > init_steps](https://github.com/mllam/neural-lam/pull/66/commits/ef3da41ac4526bc812708bb027268ceea041f73a)

Reply:
Thanks for realizing and fixing this! I think this looks correct now.

Since this is something that is both important and easy to get wrong (lots of indices and offsets being juggled around) I also added a test for this: https://github.com/mllam/neural-lam/pull/66/commits/8cc6c3dc2ff9f099f2e6ae6fcb0d77637d6b3a47 where I compared the resulting dataset length from different configs with manually computed lengths (feel free to expand on or improve this test). 

Leaving this conversation open since maybe @leifdenby wants to check, then feel free to resolve.

Reply:
with the test you added in https://github.com/mllam/neural-lam/commit/8cc6c3dc2ff9f099f2e6ae6fcb0d77637d6b3a47 @joeloskarsson this is completely clear to me now, thanks!

Reply:
Great, resolving this.